feat(getstepdefinitionspaths): configurable nonGlobalStepDefinitions base directory

liamchilds opened the pull request:
Within my team we have a preference to split the step definitions away from where the feature files are located.

However we love the ability to scope test definitions so using the common folder and global approach doesn't seem right.

To this end I have created a PR that if you agree adds a new configuration option `nonGlobalStepBaseDir` which defines a location from which the feature files folders are matched from. If it is not set then the current behaviour is maintained.

sagorla commented:
Hi @liamchilds 

looks like commonPath is stilling referring to  "cypress/integration/common" path, not switching to "nonGlobalStepBaseDir" path.

liamchilds commented:
Thanks for taking a look, that was via design if I understand your point correctly. I suppose it seems counter intuitive on reflection.

This was only to set the path which is used for the non global step configuration, however I could change it so that the common settings are also overridden.

What is your desired outcome here, am I right in thinking that if `nonGlobalStepBaseDir` is defined and commonPath is not then that location is used for both. However when commonPath is also defined it takes precedence over `nonGlobalStepBaseDir`?

sagorla commented:
@liamchilds 
currently, as per design we need to maintain  "commonPath" & "non global step definitions" in   "cypress/integration/" folder when "nonGlobalStepDefinitions" is true.

The same way we should allow to  maintain "commonPath" & "non global step definitions" in   "nonGlobalStepBaseDir" folder when "nonGlobalStepDefinitions" is true & "nonGlobalStepBaseDiri" is defined. So that all step definitions (common & global) will be in the same folder so that maintaining step definitions will be easy. This is my just personal opinion.

liamchilds commented:
@sagorla I have updated the branch with your suggestions. Common path is now maintained from nonGlobalStepBaseDir when defined.

sagorla commented:
Hi @liamchilds 

commonPath step definition file is not getting executed when nonGlobalStepBaseDir is defined. Here I have noticed you are trying to get commonPath with the reference of nonGlobalPath but this path is good for step definitions for the feature which is running. Here we will maintain only non global only .js files for that specific feature file. for common step definitions you may need to consider step base dir.

if (loaded.config.nonGlobalStepBaseDir) {
      const stepBase = `${appRoot}/${loaded.config.nonGlobalStepBaseDir}`;
      nonGlobalPath = nonGlobalPath.replace(featureBase, stepBase);
      **commonPath = `${nonGlobalPath}/${loaded.config.commonPath || "/common"}`;**
    }
correct path will be as below
    **commonPath = loaded.config.commonPath || `${stepBase}/common/`;**

lgandecki commented:
shoot, I completely zoned out about this one. Will review today or tomorrow at max. Putting on the daily todo list :)  

Do you guys think it would also solve https://github.com/TheBrainFamily/cypress-cucumber-preprocessor/pull/275  ?

lgandecki commented:
@liamchilds do you agree with @sagorla comment? Are you still using it? it'd be great to iron out the API here so we don't have to introduce breaking changes.

pmelab commented:
This would also enable us to swap out step implementations for tests. For example, using the same scenario to test on API- and UI-level. Just tested it and it works perfectly with a custom cosmic-config:

```js
module.exports = {
  nonGlobalStepDefinitions: true,
  nonGlobalStepBaseDir: process.env.CYPRESS_TEST_LEVEL === 'api'
    ? 'cypress/integration/api/'
    : 'cypress/integration/ui/'
}
```

Also, I would agree with @sagorla on the common-path issue.

liamchilds commented:
@lgandecki Apologies I was swamped with work and didnt get round to finishing this work off.
Looking over it today I agree with the comments @sagorla made and agree with what was suggested.
I have also tidied up the Readme to hopefully make it more clear about what was intended and finally I also updated with master and rebased to make the changes clear against current master.

If you could have another look and let me know what you think that would be great :)

liamchilds commented:
Oh and yes I think it does solve https://github.com/TheBrainFamily/cypress-cucumber-preprocessor/pull/275 as well

liamchilds commented:
@lgandecki Could you take another look please?

lgandecki commented:
Thanks for updating and rebasing! 
I might be missing something, but why don't we just use stepDefinitions option in the configuration? It's not like someone can have both nonGlobal and global set at the same time, to have separate stepDefinition location configuration. 
The .feature files location is AFAIR set in the cypress config itself.

lgandecki commented:
@liamchilds ^

liamchilds commented:
To be honest I've had a look and I think it probably could but I don't understand any other side effects this could cause.

The original goal of the PR was to explicitly override the location of where the library looks for the nonGlobal steps. I wanted to make the config option explicit so that in the future the stepDefinition option doesn't get overloaded.

It just depends how you want this PR to proceed.

lgandecki commented:
@liamchilds  I don't think it could even cause any side effects? for the global configuration nothing should change. And for the nonGlobal, instead of a new stepDefinition option, you use the same as you would use for global. 
What kind of side-effects are you thinking?

liamchilds commented:
I will take a look tomorrow when I can have a more thorough look through the surrounding code. It may have just been from a lack of wider understanding of how it all fits together.

lgandecki commented:
ok great. I also have some time this week to work on this, let's get this through the finish line :) quite a lot of people are asking for these changes. Thanks so much for your time and help!

jp7677 commented on lib/getStepDefinitionsPaths.js:
```diff
@@ -1,23 +1,35 @@
-    const commonPath =
-      loaded.config.commonPath || `${stepDefinitionPath()}/common/`;
+    let nonGlobalPath = getStepDefinitionPathsFrom(filePath);
+    let commonPath = loaded.config.commonPath || `${featureBase}/common/`;
```
@liamchilds  @lgandecki 
I have tried this PR to test if it resolves the issue described in https://github.com/TheBrainFamily/cypress-cucumber-preprocessor/pull/275 (step definitions defined in a custom commonPath are not found in my setup (webpack, typescript)).

It did not because the `appRoot` is only used when `nonGlobalStepBaseDir` is used. When changing this line to
```typescript
let commonPath = `${appRoot}/${loaded.config.commonPath}` || `${featureBase}/common/`;
```
the `commonPath` parameter works for me, even without specifying `nonGlobalStepBaseDir`.

lgandecki replied:
thanks for the explanation, I think I will take this PR and add the appRoot as you suggested. Just have to test it a bit.

LPCmedia commented:
I have a rebased branch actually did not get any conflicts. But yes Very relevant.

lgandecki commented:
:tada: This PR is included in version 2.5.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/cypress-cucumber-preprocessor)
- [GitHub release](https://github.com/TheBrainFamily/cypress-cucumber-preprocessor/releases/tag/v2.5.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

Eduardo-Carneiro-CM commented:
I'm trying to use the commonPath config in package.json and it still says "Step implementation missing for"

seraphilus commented:
> I'm trying to use the commonPath config in package.json and it still says "Step implementation missing for"

Seems like it is still not working with v4.0.0. 
Have you managed to find some solution to this?
